exporter/opensearchexporter/opensearch_bulk.go

```diff
@@ -0,0 +1,214 @@
+		DisableRetry:  retryDisabled,
+		MaxRetries:    maxRetries,
+		RetryBackoff:  createBackoffFunction(&config.Retry),
+
+		// configure internal metrics reporting and logging
+		EnableMetrics:     false, // TODO
+		EnableDebugLogger: false, // TODO
+		Logger:            (*clientLogger)(logger),
+	})
+}
+
+func newTransport(config *Config, tlsCfg *tls.Config) *http.Transport {
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	if tlsCfg != nil {
+		transport.TLSClientConfig = tlsCfg
+	}
+	if config.ReadBufferSize > 0 {
+		transport.ReadBufferSize = config.ReadBufferSize
+	}
+	if config.WriteBufferSize > 0 {
+		transport.WriteBufferSize = config.WriteBufferSize
+	}
+
+	return transport
+}
+
+func newBulkIndexer(logger *zap.Logger, client *opensearch.Client, config *Config) (osBulkIndexerCurrent, error) {
+	// TODO: add debug logger
+	return opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
+		NumWorkers:    config.NumWorkers,
+		FlushBytes:    config.Flush.Bytes,
+		FlushInterval: config.Flush.Interval,
+		Client:        client,
+		Pipeline:      config.Pipeline,
+		Timeout:       config.Timeout,
+
+		OnError: func(_ context.Context, err error) {
+			logger.Error(fmt.Sprintf("Bulk indexer error: %v", err))
+		},
+	})
+}
+
+func createBackoffFunction(config *RetrySettings) func(int) time.Duration {
+	if !config.Enabled {
+		return nil
+	}
+
+	expBackoff := backoff.NewExponentialBackOff()
+	if config.InitialInterval > 0 {
+		expBackoff.InitialInterval = config.InitialInterval
+	}
+	if config.MaxInterval > 0 {
+		expBackoff.MaxInterval = config.MaxInterval
+	}
+	expBackoff.Reset()
+
+	return func(attempts int) time.Duration {
+		if attempts == 1 {
+			expBackoff.Reset()
+		}
+
+		return expBackoff.NextBackOff()
+	}
+}
+
+func shouldRetryEvent(status int) bool {
+	for _, retryable := range retryOnStatus {
+		if status == retryable {
+			return true
+		}
+	}
+	return false
+}
+
+func pushDocuments(ctx context.Context, logger *zap.Logger, index string, document []byte, bulkIndexer osBulkIndexerCurrent, maxAttempts int) error {
+	attempts := 1
+	body := bytes.NewReader(document)
+	item := osBulkIndexerItem{Action: createAction, Index: index, Body: body}
+	// Setup error handler. The handler handles the per item response status based on the
+	// selective ACKing in the bulk response.
+	item.OnFailure = func(ctx context.Context, item osBulkIndexerItem, resp osBulkIndexerResponseItem, err error) {
+		switch {
+		case attempts < maxAttempts && shouldRetryEvent(resp.Status):
+			logger.Debug("Retrying to index",
+				zap.String("name", index),
+				zap.Int("attempt", attempts),
+				zap.Int("status", resp.Status),
+				zap.NamedError("reason", err))
+
+			attempts++
+			_, _ = body.Seek(0, io.SeekStart)
+			_ = bulkIndexer.Add(ctx, item)
+
+		case resp.Status == 0 && err != nil:
+			// Encoding error. We didn't even attempt to send the event
+			logger.Error("Drop docs: failed to add docs to the bulk request buffer.",
```

Comment:
I am not sure on the severity of dropped docs. Since the collector remains running and connecting to OpenSearch, I would see this with a lower severity. My thinking is, that dropping some of the docs is less severe as dropping all of them due to misconfiguration or network partition. Especially indexing errors (further down) can occur due to conflicts within a particular document.

Reply:
Does this instrumentation expose metrics on the number of dropped docs? This would really help observability of the Collector.